mantidproject/mantid
# Improve detector 2theta coverage calculation in SofQWNormalisedPolygon

The method of calculating detector's 2theta coverage by picking a point in the bounding box's center and top is not very accurate. This PR improves the calculation in several ways:
* Instead of expressing the 2theta coverage as nominal 2theta +/- width/2, we now use just the minimum and maximum 2theta. The nominal 2theta does not always correspond to the actual average 2theta.
* Check if a detector has a simple shape where the 2thetas can be calculated accurately. In case of cuboid and cylinder, we do just this.
* More complicated detector shapes will still be handled through the bounding box. However, we now pick points on each side of the box and compute the 2thetas for those.
* There is now an additional property `DetectorTwoThetaRanges` which accepts a `TableWorkspace` tabulating the per-detector min and max 2thetas. Handy for overriding the calculation in the algorithm.

**Report to:** [nobody].

**To test:**

With unit test data in Mantid's data search directories, the following script should run without problems:
```python
ws = DirectILLCollectData('ILL/IN4/084447.nxs')
ws = ConvertUnits(ws, 'DeltaE', 'Direct')
ws = Rebin(ws, [0.02])
sqw = SofQWNormalisedPolygon(ws, [0., 0.04, 4.], 'Direct')
```

Observe the gap (NaNs) between the low angle (small Q, around zero energy transfer) and large angle detectors.

To test the new 2theta table input property, this file can be used:
[2theta_table.zip](https://github.com/mantidproject/mantid/files/2732471/2theta_table.zip)
The script changes to: 
```python
ws = DirectILLCollectData('ILL/IN4/084447.nxs')
ws = ConvertUnits(ws, 'DeltaE', 'Direct')
ws = Rebin(ws, [0.02])
sqw = SofQWNormalisedPolygon(ws, [0., 0.04, 4.], 'Direct', DetectorTwoThetaRanges='2theta')
```

The 2thetas in the above file are calculated using an analytical formula from the instrument scientists --- the results should be close to identical when not using the pre-calcualted table *except* for the small angle detectors (low Q) for which the analytical formula is wrong.


Fixes #24305.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
